scapy/contrib/wireguard.py

```diff
@@ -0,0 +1,119 @@
+# This file is part of Scapy
+# Scapy is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# any later version.
+#
+# Scapy is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Scapy. If not, see <http://www.gnu.org/licenses/>.
+
+
+# scapy.contrib.description = WireGuard
+# scapy.contrib.status = loads
+
+"""WireGuard Module
+Implements the WireGuard network tunnel protocol.
+Based on the whitepaper: https://www.wireguard.com/papers/wireguard.pdf
+"""
+
+from scapy.fields import ByteEnumField, ThreeBytesField, XIntField, \
+    XStrFixedLenField, LongField, XStrField
+from scapy.layers.inet import UDP
+from scapy.packet import Packet, bind_layers
+
+
+class Wireguard(Packet):
+    """
+    Wrapper that only contains the message type.
+    """
+    name = "Wireguard"
+
+    fields_desc = [
+        ByteEnumField(
+            "type", 1,
+            {
+                1: "initiate",
+                2: "respond",
+                3: "cookie reply",
+                4: "transport"
+            }
+        ),
+        ThreeBytesField("reserved", 0)
+    ]
+
+    def guess_payload_class(self, payload):
+        wireguard_types = {
+            1: WireguardInitiation,
+            2: WireguardResponse,
+            3: WireguardCookieReply,
+            4: WireguardTransport
+        }
+
+        try:
+            return wireguard_types[self.type]
+        except KeyError:
+            return Packet.guess_payload_class(self, payload)
+
+
+class WireguardInitiation(Packet):
+    name = "Wireguard Initiation"
+
+    overload_fields = {Wireguard: {"type": 1}}
+
+    fields_desc = [
+        XIntField("sender", 0),
+        XStrFixedLenField("ephemeral", 0, 32),
+        XStrFixedLenField("static", 0, 48),
+        XStrFixedLenField("tstamp", 0, 28),
+        XStrFixedLenField("mac1", 0, 16),
+        XStrFixedLenField("mac2", 0, 16),
```

Comment:
When we build new layers, best practice is to keep the names from the official doc:
```
msg = handshake_initiation {
    u8 message_type
    u8 reserved_zero[3]
    u32 sender_index
    u8 unencrypted_ephemeral[32]
    u8 encrypted_static[AEAD_LEN(32)]
    u8 encrypted_timestamp[AEAD_LEN(12)]
    u8 mac1[16]
    u8 mac2[16]
}
```
Correct me If I'm wrong.
(I know this isn't the case in Scapy)